live-tests/tests/common/mod.rs

```diff
@@ -0,0 +1,248 @@
+        "postgresql://root@{}/omicron?sslmode=disable",
+        sockaddrs
+            .into_iter()
+            .map(|a| a.to_string())
+            .collect::<Vec<_>>()
+            .join(",")
+    )
+    .parse::<PostgresConfigWithUrl>()
+    .context("failed to parse constructed postgres URL")?;
+
+    let db_config = nexus_db_queries::db::Config { url };
+    let pool = Arc::new(nexus_db_queries::db::Pool::new(log, &db_config));
+    DataStore::new_failfast(log, pool)
+        .await
+        .context("creating DataStore")
+        .map(Arc::new)
+}
+
+/// Performs quick checks to determine if the user is running these tests in the
+/// wrong place and bails out if so
+///
+/// This isn't perfect but seeks to fail fast in obviously bogus environments
+/// that someone might accidentally try to run this in.
+async fn check_execution_environment(
+    resolver: &Resolver,
+) -> Result<(), anyhow::Error> {
+    ensure!(
+        cfg!(target_os = "illumos"),
+        "live tests can only be run on deployed systems, which run illumos"
+    );
+
+    // The only real requirement for these tests is that they're run from a
+    // place with connectivity to the underlay network of a deployed control
+    // plane.  The easiest way to tell is to look up something in internal DNS.
+    resolver.lookup_ip(ServiceName::InternalDns).await.map_err(|e| {
+        let text = format!(
+            "check_execution_environment(): failed to look up internal DNS \
+                 in the internal DNS servers.\n\n \
+                 Are you trying to run this in a development environment?  \
+                 This test can only be run on deployed systems and only from a \
+                 context with connectivity to the underlay network.\n\n \
+                 raw error: {}",
+            slog_error_chain::InlineErrorChain::new(&e)
+        );
+        anyhow!("{}", textwrap::wrap(&text, 80).join("\n"))
+    })?;
+
+    // Warn the user if the temporary directory is /tmp.  This check is
+    // heuristic.  There are other ways they may have specified a tmpfs
+    // temporary directory and we don't claim to catch all of them.
+    //
+    // We could also just go ahead and use /var/tmp, but it's not clear we can
+    // reliably do that at this point (if Rust or other components have cached
+    // TMPDIR) and it would be hard to override.
+    let tmpdir = std::env::temp_dir();
+    let mut tmpdir_components = tmpdir.components().take(2);
+    if let Some(first) = tmpdir_components.next() {
+        if let Some(next) = tmpdir_components.next() {
+            if first == Component::RootDir
+                && next == Component::Normal(OsStr::new("tmp"))
+            {
+                eprintln!(
+                    "WARNING: temporary directory appears to be under /tmp, \
+                     which is generally tmpfs.  Consider setting \
+                     TMPDIR=/var/tmp to avoid runaway tests using too much\
+                     memory and swap."
```

Comment:
Should we check for a minimum amount of free space required for `TMPDIR`?

Reply:
I don't think so.  Management of tmp space seems outside the scope of all this.  I almost didn't bother with this warning at all except that the failure mode of running out of space in /tmp (i.e., swap) is usually pretty bad for the whole zone, and that could be pretty annoying to clean up in the switch zone.  That said, I'm barely pro for even this warning -- I could be convinced to just remove this altogether.

Also: there are so many ways I've seen that kind of check be wrong, in part because it's _really_ hard to know how much physical disk space something is going to need before it writes it.  Example: false positive errors when using zfs with compression=on because a thing thinks it needs a GiB even though after compression it'll only use 150 MiB of space.  Plus other things can use space in the meantime or space can free up in the meantime.